controllers/upgrade.js

```diff
@@ -0,0 +1,76 @@
+const PouchDB = require('pouchdb-core');
+PouchDB.plugin(require('pouchdb-adapter-http'));
+
+const buildDbUrl = 'https://staging.dev.medicmobile.org/_couch/builds';
+
+const buildDb = new PouchDB(buildDbUrl);
+const targetDbUrl = process.env.COUCH_URL;
+const targetDb = new PouchDB(targetDbUrl);
+
+const stagedDdocName = name => name + ':staged';
+
+const ddocName = buildInfo =>
+ `${buildInfo.namespace}:${buildInfo.application}:${buildInfo.version}`;
+
+module.exports = (buildInfo, username) => {
+  if (!buildInfo) {
+    throw new Error('Bad request');
+  }
+
+  // While we've set the api to in theory be more generic, this is for if we
+  // split this code out (and say, put it with horticulturalist). For now we
+  // just support medic deploying medic :-)
+  if (buildInfo.namespace !== 'medic' ||
+      buildInfo.application !== 'medic') {
+    return Promise.reject({
+      expected: true,
+      status: 400,
+      message: 'We only support medic as the application and namespace'
+    });
+  }
+
+  console.log('upgrade()', `Upgrading to ${JSON.stringify(buildInfo)}…`);
+
+  console.log('upgrade()', 'Fetching newDdoc…');
+  return buildDb
+    .get(ddocName(buildInfo), { attachments:true })
+    .catch(err => {
+      console.log('GOT ERROR', err);
+      if (err.status === 404) {
+        err = new Error(`Version not found: ${buildInfo.version}`);
+        err.expected = true;
+      }
+      // FIXME: why does this from causing an unhandled promise rejection?
+      //        It shouldn't do this because we catch again in server.js:194
+      //        Exceptions thrown in other parts (such as targetDb.put(newDDoc))
+      //        correctly flow to the second catch, but this one doesn't?
+      throw err;
+    })
+    .then(newDdoc => {
+      console.log('upgrade()', 'Fetched newDdoc.');
+
+      console.log('upgrade()', 'Fetching oldDdoc…');
+      return targetDb.get('_design/medic')
+        .then(oldDdoc => {
+          console.log('upgrade()', 'Fetched oldDdoc.');
+
+          newDdoc._id = stagedDdocName(oldDdoc._id);
+          delete newDdoc._rev;
+
+          newDdoc.deploy_info = {
```

Comment:
The pros for me are organizational and separation of concerns/cleanliness.  I don't think deployment data has anything to do with the app itself.  The app (ddoc) should already have all the information we need on it when it gets built/released like commit hash and build time, and after that it never gets modified (app_settings is the only exception atm but is also a hack that probably needs to get fixed eventually).  All the meta data we need about the app/ddoc becomes part of the published asset and deployment data is managed separately.  This change seems to muddy the waters.  The ddoc to me feels like it should be treated like a binary, so modifying it when we install it seems weird, I'm hesitant.